JPI2 Plugin Attempts to Resolve `defaultRuntime` Without an Exclusive Lock

I am working on moving an existing internal Jenkins plugin from using the JPI plugin to using the JPI2 plugin and have run into an issue.

When running with Gradle 9.x (tested with 9.0.0, 9.2.1, 9.6.0, and 9.7.0; not reproducible with 8.14.5) and the Artifactory plugin for Gradle, the `ManifestAction`'s `execute` method attempts to resolve the `defaultRuntime` configuration without an exclusive lock, which raises an exception within Gradle.

<details>
<summary>Stack trace</summary>

```
$ ./gradlew --stacktrace check

FAILURE: Build failed with an exception.

* What went wrong:
Failed to calculate the value of property 'componentArtifacts'.
> Could not create task ':jar'.
   > Resolution of the configuration ':defaultRuntime' was attempted without an exclusive lock. This is unsafe and not allowed.

* Try:
> For more information, please refer to https://docs.gradle.org/9.6.0/userguide/viewing_debugging_dependencies.html#sub:resolving-unsafe-configuration-resolution-errors in the Gradle documentation.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights from a Build Scan (powered by Develocity).
> Get more help at https://help.gradle.org.

* Exception is:
org.gradle.api.internal.provider.AbstractProperty$PropertyQueryException: Failed to calculate the value of property 'componentArtifacts'.
        at org.gradle.api.internal.provider.AbstractProperty.finalizeNow(AbstractProperty.java:315)
        at org.gradle.api.internal.provider.AbstractProperty.lambda$beforeRead$0(AbstractProperty.java:307)
        at org.gradle.api.internal.provider.ValueState.finalizeOnReadIfNeeded(ValueState.java:142)
        at org.gradle.api.internal.provider.AbstractProperty.beforeRead(AbstractProperty.java:307)
        at org.gradle.api.internal.provider.AbstractProperty.beforeRead(AbstractProperty.java:299)
        at org.gradle.api.internal.provider.AbstractProperty.calculatePresence(AbstractProperty.java:85)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.isPresent(AbstractMinimalProvider.java:90)
        at org.gradle.api.publish.maven.internal.publication.DefaultMavenPublication.populateFromComponent(DefaultMavenPublication.java:263)
        at org.gradle.api.publish.maven.internal.publication.DefaultMavenPublication.getArtifacts(DefaultMavenPublication.java:282)
        at org.gradle.api.publish.maven.internal.publication.DefaultMavenPublication_Decorated.getArtifacts(Unknown Source)
        at org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask.createDependencyOnMavenPublications(ArtifactoryTask.java:338)
        at org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask.checkDependsOnArtifactsToPublish(ArtifactoryTask.java:210)
        at org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask.publications(ArtifactoryTask.java:179)
        at Build_gradle$3$1$2.execute(build.gradle.kts:39)
        at Build_gradle$3$1$2.execute(build.gradle.kts:39)
        at org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask.evaluateTask(ArtifactoryTask.java:117)
        at org.jfrog.gradle.plugin.artifactory.listener.ProjectsEvaluatedBuildListener.evaluate(ProjectsEvaluatedBuildListener.java:60)
        at org.jfrog.gradle.plugin.artifactory.listener.ProjectsEvaluatedBuildListener.lambda$projectsEvaluated$3(ProjectsEvaluatedBuildListener.java:133)
        at org.jfrog.gradle.plugin.artifactory.listener.ProjectsEvaluatedBuildListener.projectsEvaluated(ProjectsEvaluatedBuildListener.java:130)
        at org.gradle.internal.code.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:124)
        at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction$1.run(DefaultListenerBuildOperationDecorator.java:173)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)
        at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingAction.execute(DefaultListenerBuildOperationDecorator.java:170)
        at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:101)
        at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:89)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:44)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:270)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:172)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:84)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:70)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:382)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:274)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:160)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:88)
        at jdk.proxy1/jdk.proxy1.$Proxy36.projectsEvaluated(Unknown Source)
        at org.gradle.initialization.ProjectsEvaluatedNotifier$NotifyProjectsEvaluatedListeners.run(ProjectsEvaluatedNotifier.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)
        at org.gradle.initialization.ProjectsEvaluatedNotifier.notify(ProjectsEvaluatedNotifier.java:35)
        at org.gradle.configuration.DefaultProjectsPreparer.prepareProjects(DefaultProjectsPreparer.java:54)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.prepareProjects(BuildTreePreparingProjectsPreparer.java:65)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild.run(BuildOperationFiringProjectsPreparer.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer.prepareProjects(BuildOperationFiringProjectsPreparer.java:40)
        at org.gradle.initialization.VintageBuildModelController.lambda$prepareProjects$2(VintageBuildModelController.java:83)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$15(StateTransitionController.java:293)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:304)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:292)
        at org.gradle.internal.model.StateTransitionController.lambda$transitionIfNotPreviously$12(StateTransitionController.java:251)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:35)
        at org.gradle.internal.model.StateTransitionController.transitionIfNotPreviously(StateTransitionController.java:247)
        at org.gradle.initialization.VintageBuildModelController.prepareProjects(VintageBuildModelController.java:83)
        at org.gradle.initialization.VintageBuildModelController.prepareToScheduleTasks(VintageBuildModelController.java:70)
        at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$prepareToScheduleTasks$7(DefaultBuildLifecycleController.java:185)
        at org.gradle.internal.model.StateTransitionController.lambda$doTransition$15(StateTransitionController.java:293)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:304)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:292)
        at org.gradle.internal.model.StateTransitionController.lambda$maybeTransition$10(StateTransitionController.java:228)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:35)
        at org.gradle.internal.model.StateTransitionController.maybeTransition(StateTransitionController.java:224)
        at org.gradle.internal.build.DefaultBuildLifecycleController.prepareToScheduleTasks(DefaultBuildLifecycleController.java:183)
        at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.scheduleRequestedTasks(DefaultBuildTreeWorkPreparer.java:35)
        at org.gradle.internal.cc.impl.barrier.BarrierAwareBuildTreeWorkPreparer.scheduleRequestedTasks$lambda$0(BarrierAwareBuildTreeWorkPreparer.kt:34)
        at org.gradle.internal.cc.impl.barrier.VintageConfigurationTimeActionRunner.runConfigurationTimeAction(VintageConfigurationTimeActionRunner.kt:48)
        at org.gradle.internal.cc.impl.barrier.BarrierAwareBuildTreeWorkPreparer.scheduleRequestedTasks(BarrierAwareBuildTreeWorkPreparer.kt:33)
        at org.gradle.internal.cc.impl.VintageBuildTreeWorkController$scheduleAndRunRequestedTasks$1$finalizedGraph$1.call(VintageBuildTreeWorkController.kt:37)
        at org.gradle.internal.cc.impl.VintageBuildTreeWorkController$scheduleAndRunRequestedTasks$1$finalizedGraph$1.call(VintageBuildTreeWorkController.kt:37)
        at org.gradle.internal.Try.ofFailable(Try.java:46)
        at org.gradle.internal.cc.impl.VintageBuildTreeWorkController$scheduleAndRunRequestedTasks$1.apply(VintageBuildTreeWorkController.kt:37)
        at org.gradle.internal.cc.impl.VintageBuildTreeWorkController$scheduleAndRunRequestedTasks$1.apply(VintageBuildTreeWorkController.kt:36)
        at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph.withNewWorkGraph(DefaultIncludedBuildTaskGraph.java:115)
        at org.gradle.internal.cc.impl.VintageBuildTreeWorkController.scheduleAndRunRequestedTasks(VintageBuildTreeWorkController.kt:36)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$scheduleAndRunTasks$1(DefaultBuildTreeLifecycleController.java:76)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$runBuild$5(DefaultBuildTreeLifecycleController.java:121)
        at org.gradle.internal.model.StateTransitionController.lambda$transition$7(StateTransitionController.java:207)
        at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:304)
        at org.gradle.internal.model.StateTransitionController.lambda$transition$8(StateTransitionController.java:207)
        at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:45)
        at org.gradle.internal.model.StateTransitionController.transition(StateTransitionController.java:207)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.runBuild(DefaultBuildTreeLifecycleController.java:118)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:76)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:71)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.internal.buildtree.ProblemReportingBuildActionRunner.run(ProblemReportingBuildActionRunner.java:55)
        at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:83)
        at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:135)
        at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:64)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.lambda$execute$0(RootBuildLifecycleBuildActionExecutor.java:101)
        at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:119)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:101)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor.runBuildTreeLifecycle(DefaultBuildTreeActionExecutor.java:126)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor.access$100(DefaultBuildTreeActionExecutor.java:52)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor$2.call(DefaultBuildTreeActionExecutor.java:98)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor$2.call(DefaultBuildTreeActionExecutor.java:94)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:210)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:205)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:54)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor.runAsBuildOperation(DefaultBuildTreeActionExecutor.java:94)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor.lambda$runBuildTreeAction$0(DefaultBuildTreeActionExecutor.java:88)
        at org.gradle.internal.work.DefaultWorkerLeaseService.lambda$runAndReleaseLocks$0(DefaultWorkerLeaseService.java:300)
        at org.gradle.internal.work.ResourceLockStatistics$1.measure(ResourceLockStatistics.java:43)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAndReleaseLocks(DefaultWorkerLeaseService.java:298)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocksAcquired(DefaultWorkerLeaseService.java:294)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:286)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:130)
        at org.gradle.launcher.exec.DefaultBuildTreeActionExecutor.runBuildTreeAction(DefaultBuildTreeActionExecutor.java:88)
        at org.gradle.tooling.internal.provider.continuous.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:111)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
        at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
        at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor$ActionImpl.apply(BuildSessionLifecycleBuildActionExecutor.java:106)
        at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor$ActionImpl.apply(BuildSessionLifecycleBuildActionExecutor.java:94)
        at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:73)
        at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor.execute(BuildSessionLifecycleBuildActionExecutor.java:67)
        at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor.execute(BuildSessionLifecycleBuildActionExecutor.java:45)
        at org.gradle.internal.buildprocess.execution.StartParamsValidatingActionExecutor.execute(StartParamsValidatingActionExecutor.java:57)
        at org.gradle.internal.buildprocess.execution.StartParamsValidatingActionExecutor.execute(StartParamsValidatingActionExecutor.java:32)
        at org.gradle.internal.buildprocess.execution.SessionFailureReportingActionExecutor.execute(SessionFailureReportingActionExecutor.java:51)
        at org.gradle.internal.buildprocess.execution.SessionFailureReportingActionExecutor.execute(SessionFailureReportingActionExecutor.java:39)
        at org.gradle.internal.buildprocess.execution.SetupLoggingActionExecutor.execute(SetupLoggingActionExecutor.java:47)
        at org.gradle.internal.buildprocess.execution.SetupLoggingActionExecutor.execute(SetupLoggingActionExecutor.java:31)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:70)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.lambda$execute$0(ForwardClientInput.java:40)
        at org.gradle.internal.daemon.clientinput.ClientInputForwarder.forwardInput(ClientInputForwarder.java:80)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:64)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ApplyClientEnvironmentVariables.doBuild(ApplyClientEnvironmentVariables.java:80)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:74)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator.lambda$runCommand$0(DaemonStateCoordinator.java:321)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
Caused by: org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreationException: Could not create task ':jar'.
        at org.gradle.api.internal.tasks.DefaultTaskContainer.taskCreationException(DefaultTaskContainer.java:778)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.access$600(DefaultTaskContainer.java:77)
        at org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreatingProvider.domainObjectCreationException(DefaultTaskContainer.java:770)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.tryCreate(DefaultNamedDomainObjectCollection.java:1016)
        at org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreatingProvider.access$1401(DefaultTaskContainer.java:717)
        at org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreatingProvider$1.run(DefaultTaskContainer.java:743)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)
        at org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreatingProvider.tryCreate(DefaultTaskContainer.java:739)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.calculateOwnValue(DefaultNamedDomainObjectCollection.java:990)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:80)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:100)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.get(DefaultNamedDomainObjectCollection.java:976)
        at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getDelegate(LazyPublishArtifact.java:87)
        at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getType(LazyPublishArtifact.java:67)
        at org.gradle.api.plugins.internal.JavaConfigurationVariantMapping$UnpublishableArtifactTypeSpec.isSatisfiedBy(JavaConfigurationVariantMapping.java:81)
        at org.gradle.api.plugins.internal.JavaConfigurationVariantMapping.execute(JavaConfigurationVariantMapping.java:60)
        at org.gradle.api.plugins.internal.JavaConfigurationVariantMapping.execute(JavaConfigurationVariantMapping.java:31)
        at org.gradle.api.publish.internal.component.ConfigurationVariantMapping.visitVariant(ConfigurationVariantMapping.java:93)
        at org.gradle.api.publish.internal.component.ConfigurationVariantMapping.collectVariants(ConfigurationVariantMapping.java:76)
        at org.gradle.api.publish.internal.component.DefaultAdhocSoftwareComponent.computeVariants(DefaultAdhocSoftwareComponent.java:116)
        at org.gradle.api.publish.internal.component.DefaultAdhocSoftwareComponent.getUsages(DefaultAdhocSoftwareComponent.java:92)
        at org.gradle.jvm.component.internal.DefaultJvmSoftwareComponent_Decorated.getUsages(Unknown Source)
        at org.gradle.api.publish.maven.internal.publication.MavenComponentParser.createSortedVariantsStream(MavenComponentParser.java:265)
        at org.gradle.api.publish.maven.internal.publication.MavenComponentParser.parseArtifacts(MavenComponentParser.java:128)
        at org.gradle.api.internal.provider.ValueSupplier$Present.transform(ValueSupplier.java:520)
        at org.gradle.api.internal.provider.TransformBackedProvider.mapValue(TransformBackedProvider.java:91)
        at org.gradle.api.internal.provider.TransformBackedProvider.calculateOwnValue(TransformBackedProvider.java:82)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:115)
        at org.gradle.api.internal.provider.Collectors$ElementsFromCollectionProvider.collectEntries(Collectors.java:239)
        at org.gradle.api.internal.provider.AbstractCollectionProperty$CollectingSupplier.lambda$calculateOwnValue$0(AbstractCollectionProperty.java:507)
        at org.gradle.api.internal.provider.AbstractCollectingSupplier.calculateValue(AbstractCollectingSupplier.java:87)
        at org.gradle.api.internal.provider.AbstractCollectionProperty$CollectingSupplier.calculateOwnValue(AbstractCollectionProperty.java:506)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:115)
        at org.gradle.api.internal.provider.AbstractCollectingSupplier.calculateValue(AbstractCollectingSupplier.java:41)
        at org.gradle.api.internal.provider.AbstractCollectionProperty.finalValue(AbstractCollectionProperty.java:318)
        at org.gradle.api.internal.provider.AbstractCollectionProperty.finalValue(AbstractCollectionProperty.java:81)
        at org.gradle.api.internal.provider.AbstractProperty.finalizeNow(AbstractProperty.java:312)
        ... 170 more
Caused by: org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$IllegalResolutionException: Resolution of the configuration ':defaultRuntime' was attempted without an exclusive lock. This is unsafe and not allowed.
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveGraphIfRequired(DefaultConfiguration.java:642)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ResolverResultsResolutionResultProvider.getValue(DefaultConfiguration.java:626)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ResolverResultsResolutionResultProvider.getValue(DefaultConfiguration.java:612)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.getGraph(DefaultResolutionResult.java:77)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.lambda$getRootComponent$0(DefaultResolutionResult.java:62)
        at org.gradle.api.internal.provider.DefaultProvider.calculateOwnValue(DefaultProvider.java:73)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:80)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:100)
        at org.gradle.api.internal.artifacts.result.DefaultResolutionResult.getRoot(DefaultResolutionResult.java:56)
        at org.jenkinsci.gradle.plugins.jpi2.ManifestAction.execute(ManifestAction.java:37)
        at org.jenkinsci.gradle.plugins.jpi2.ManifestAction.execute(ManifestAction.java:18)
        at org.gradle.jvm.tasks.Jar.manifest(Jar.java:208)
        at org.gradle.api.tasks.bundling.Jar_Decorated.manifest(Unknown Source)
        at org.jenkinsci.gradle.plugins.jpi2.V2JpiPlugin$3.execute(V2JpiPlugin.java:152)
        at org.jenkinsci.gradle.plugins.jpi2.V2JpiPlugin$3.execute(V2JpiPlugin.java:149)
        at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:66)
        at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:66)
        at org.gradle.internal.code.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:124)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:109)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:106)
        at org.gradle.internal.ImmutableActionSet$SetWithManyActions.execute(ImmutableActionSet.java:330)
        at org.gradle.api.internal.DefaultDomainObjectCollection.doAdd(DefaultDomainObjectCollection.java:283)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection.doAdd(DefaultNamedDomainObjectCollection.java:126)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.tryCreate(DefaultNamedDomainObjectCollection.java:1010)
        ... 209 more
```
</details>

See [the replication repository](https://github.com/mtughan/jpi-plugin-tests/tree/artifactory) for more details. In our case, we are publishing to an internal Artifactory server, but using the Jenkins project's Artifactory server is able to replicate the issue.

I am unsure if this is specific to the Artifactory plugin or if others can also come across this, but this is what I found.

--- a/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/ConfigureJpiAction.java
+++ b/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/ConfigureJpiAction.java
@@ -2,6 +2,7 @@ package org.jenkinsci.gradle.plugins.jpi2;
 
 import org.gradle.api.Action;
 import org.gradle.api.Project;
+import org.gradle.api.Task;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.file.CopySpec;
 import org.gradle.api.tasks.bundling.War;
@@ -29,7 +30,23 @@ class ConfigureJpiAction implements Action<War> {
     @Override
     public void execute(@NotNull War jpi) {
         jpi.getArchiveExtension().set(extension.getArchiveExtension());
-        jpi.manifest(new ManifestAction(project, configuration, extension));
+        jpi.manifest(new ManifestAction(project, extension));
+
+        // Resolving `configuration` must wait until the jpi task actually executes: some publishing
+        // plugins (e.g. com.jfrog.artifactory) realize this task from a gradle.projectsEvaluated
+        // listener, before projects are configured and before Gradle's exclusive project-execution
+        // lock is available, and an eager resolution there is rejected as unsafe.
+        var pluginDependencies = project.provider(() -> V2JpiPlugin.resolvePluginDependencies(configuration));
+        jpi.getInputs().property("pluginDependencies", pluginDependencies).optional(true);
+        jpi.doFirst(new Action<>() {
+            @Override
+            public void execute(@NotNull Task task) {
+                var value = pluginDependencies.getOrNull();
+                if (value != null) {
+                    jpi.getManifest().getAttributes().put("Plugin-Dependencies", value);
+                }
+            }
+        });
         jpi.from(project.getTasks().named("jar"), new Action<>() {
             @Override
             public void execute(@NotNull CopySpec copySpec) {

--- a/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/ManifestAction.java
+++ b/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/ManifestAction.java
@@ -2,28 +2,25 @@ package org.jenkinsci.gradle.plugins.jpi2;
 
 import org.gradle.api.Action;
 import org.gradle.api.Project;
-import org.gradle.api.artifacts.Configuration;
-import org.gradle.api.internal.artifacts.result.DefaultResolvedDependencyResult;
 import org.gradle.api.java.archives.Manifest;
 import org.gradle.api.plugins.JavaPluginExtension;
 import org.gradle.jvm.toolchain.JavaLanguageVersion;
 import org.jetbrains.annotations.NotNull;
 
-import java.util.Objects;
 import java.util.stream.Collectors;
 
 /**
- * Action to update the JAR manifest with plugin dependencies and other attributes required in a Jenkins Plugin.
+ * Action to update the JAR manifest with attributes required in a Jenkins Plugin, except for
+ * {@code Plugin-Dependencies}, which requires resolving a configuration and is set separately at task
+ * execution time (see {@link V2JpiPlugin}).
  */
 class ManifestAction implements Action<Manifest> {
     public static final int DEFAULT_MINIMUM_JAVA_VERSION = 17;
     private final Project project;
-    private final Configuration configuration;
     private final JenkinsPluginExtension extension;
 
-    public ManifestAction(Project project, Configuration configuration, JenkinsPluginExtension extension) {
+    public ManifestAction(Project project, JenkinsPluginExtension extension) {
         this.project = project;
-        this.configuration = configuration;
         this.extension = extension;
     }
 
@@ -33,23 +30,6 @@ class ManifestAction implements Action<Manifest> {
         var version = extension.getEffectiveVersion().get();
         attributes.put("Implementation-Title", project.getGroup() + "#" + project.getName() + ";" + version);
         attributes.put("Implementation-Version", version);
-
-        var rootDependencies = configuration.getIncoming().getResolutionResult().getRoot().getDependencies();
-
-        var pluginDependencies = rootDependencies
-                .stream()
-                .filter(it -> !it.isConstraint())
-                .filter(it -> it instanceof DefaultResolvedDependencyResult)
-                .map(it -> ((DefaultResolvedDependencyResult) it))
-                .filter(it -> it.getResolvedVariant().getDisplayName().equals(HpiMetadataRule.DEFAULT_RUNTIME_VARIANT))
-                .map(it -> it.getSelected().getModuleVersion())
-                .filter(Objects::nonNull)
-                .map(it -> it.getName() + ":" + it.getVersion())
-                .toList();
-
-        if (!pluginDependencies.isEmpty()) {
-            attributes.put("Plugin-Dependencies", String.join(",", pluginDependencies));
-        }
         attributes.put("Plugin-Version", version);
         attributes.put("Short-Name", extension.getPluginId().get());
         attributes.put("Extension-Name", extension.getPluginId().get());

--- a/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/V2JpiPlugin.java
+++ b/jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/V2JpiPlugin.java
@@ -11,6 +11,7 @@ import org.gradle.api.artifacts.ResolvedArtifact;
 import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
 import org.gradle.api.attributes.Usage;
 import org.gradle.api.file.Directory;
+import org.gradle.api.internal.artifacts.result.DefaultResolvedDependencyResult;
 import org.gradle.api.plugins.BasePlugin;
 import org.gradle.api.plugins.ExtensionAware;
 import org.gradle.api.plugins.GroovyBasePlugin;
@@ -34,6 +35,7 @@ import org.jenkinsci.gradle.plugins.jpi2.localization.LocalizationPlugin;
 import org.jenkinsci.gradle.plugins.jpi2.accmod.CheckAccessModifierTask;
 import org.jenkinsci.gradle.plugins.jpi2.accmod.PrefixedPropertiesProvider;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -45,6 +47,7 @@ import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Collectors;
 
@@ -169,11 +172,27 @@ public class V2JpiPlugin implements Plugin<Project> {
         project.getTasks().named("jar", Jar.class).configure(new Action<>() {
             @Override
             public void execute(@NotNull Jar jarTask) {
-                jarTask.manifest(new ManifestAction(project, defaultRuntime, extension));
+                jarTask.manifest(new ManifestAction(project, extension));
                 jarTask.getInputs().file(optionalManifestFile);
                 jarTask.getManifest().from(optionalManifestFile);
                 jarTask.setPreserveFileTimestamps(false);
                 jarTask.setReproducibleFileOrder(true);
+
+                // Resolving defaultRuntime must wait until the jar task actually executes: some publishing
+                // plugins (e.g. com.jfrog.artifactory) realize this task from a gradle.projectsEvaluated
+                // listener, before projects are configured and before Gradle's exclusive project-execution
+                // lock is available, and an eager resolution there is rejected as unsafe.
+                var pluginDependencies = project.provider(() -> resolvePluginDependencies(defaultRuntime));
+                jarTask.getInputs().property("pluginDependencies", pluginDependencies).optional(true);
+                jarTask.doFirst(new Action<>() {
+                    @Override
+                    public void execute(@NotNull Task task) {
+                        var value = pluginDependencies.getOrNull();
+                        if (value != null) {
+                            jarTask.getManifest().getAttributes().put("Plugin-Dependencies", value);
+                        }
+                    }
+                });
             }
         });
         Provider<Directory> jpiDirectory = project.getLayout().getBuildDirectory().dir("jpi");
@@ -622,6 +641,24 @@ public class V2JpiPlugin implements Plugin<Project> {
     private static String getVersionFromProperties(@NotNull Project project, String propertyName, String defaultVersion) {
         Provider<String> myProperty = project.getProviders().gradleProperty(propertyName);
         return myProperty.getOrElse(defaultVersion);
+    }
+
+    @Nullable
+    static String resolvePluginDependencies(@NotNull Configuration defaultRuntime) {
+        var rootDependencies = defaultRuntime.getIncoming().getResolutionResult().getRoot().getDependencies();
+
+        var pluginDependencies = rootDependencies
+                .stream()
+                .filter(it -> !it.isConstraint())
+                .filter(it -> it instanceof DefaultResolvedDependencyResult)
+                .map(it -> ((DefaultResolvedDependencyResult) it))
+                .filter(it -> it.getResolvedVariant().getDisplayName().equals(HpiMetadataRule.DEFAULT_RUNTIME_VARIANT))
+                .map(it -> it.getSelected().getModuleVersion())
+                .filter(Objects::nonNull)
+                .map(it -> it.getName() + ":" + it.getVersion())
+                .toList();
+
+        return pluginDependencies.isEmpty() ? null : String.join(",", pluginDependencies);
     }
 
 }

--- a/jpi2/src/test/java/org/jenkinsci/gradle/plugins/jpi2/ManifestIntegrationTest.java
+++ b/jpi2/src/test/java/org/jenkinsci/gradle/plugins/jpi2/ManifestIntegrationTest.java
@@ -18,6 +18,7 @@ import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
 
 class ManifestIntegrationTest extends V2IntegrationTestBase {
 
@@ -188,6 +189,22 @@ class ManifestIntegrationTest extends V2IntegrationTestBase {
         ith.gradleRunner().withArguments("build").build();
 
         assertThat(manifestAttributes(ith).getValue("Support-Dynamic-Loading")).isEqualTo("false");
+    }
+
+    @Test
+    void manifestCanBeComputedWhenPublicationArtifactsAreAccessedFromProjectsEvaluated() throws IOException {
+        // Publishing plugins such as com.jfrog.artifactory inspect every maven publication's
+        // artifacts from a gradle.projectsEvaluated listener, which realizes the jar task (and its
+        // manifest) outside of the per-project configuration lock that plain afterEvaluate holds.
+        var ith = new IntegrationTestHelper(tempDir, "8.14");
+        initBuild(ith);
+        Files.write(ith.inProjectDir("build.gradle.kts").toPath(), (getBasePluginConfig() + /* language=kotlin */ """
+                gradle.projectsEvaluated {
+                    (publishing.publications["mavenJpi"] as MavenPublication).artifacts.forEach { }
+                }
+                """).getBytes(StandardCharsets.UTF_8));
+
+        assertThatCode(() -> ith.gradleRunner().withArguments("help").build()).doesNotThrowAnyException();
     }
 
     private static Attributes manifestAttributes(IntegrationTestHelper ith) throws IOException {

